File: tests/DesktopGrass.Win2D.Tests/SimTests/DesertTests.cs
// DesertTests.cs - §14 Desert scene cacti + tumbleweeds.

using System;
using System.Linq;
using DesktopGrass.Win2D;
using Xunit;

namespace DesktopGrass.Win2D.Tests.SimTests;

public class DesertTests
{
    private const double Monitor1920 = 1920.0;

    private readonly struct ExpectedCactus
    {
        public readonly int SlotIndex;
        public readonly byte Type;
        public readonly double Height;
        public readonly double Width;
        public readonly sbyte ArmSide;

        public ExpectedCactus(int slotIndex, byte type, double height, double width, sbyte armSide)
        {
            SlotIndex = slotIndex;
            Type = type;
            Height = height;
            Width = width;
            ArmSide = armSide;
        }
    }

    private static Sim BuildSim(double density = Constants.DEFAULT_DENSITY)
    {
        var sim = new Sim
        {
            Blades = Sim.GenerateBlades(Constants.CANONICAL_TEST_SEED, Monitor1920, density),
            WindowHeight = Constants.STRIP_HEIGHT + Constants.HEADROOM,
            GroundY = Constants.STRIP_HEIGHT + Constants.HEADROOM,
        };
        sim.ResetAmbientGusts(Constants.CANONICAL_TEST_SEED, Monitor1920);
        sim.ResetEntities(Constants.CANONICAL_TEST_SEED);
        return sim;
    }

    private static ExpectedCactus FirstExpectedCactus(int bladeCount)
    {
        var prng = Prng.Init(Constants.CANONICAL_TEST_SEED ^ Constants.CACTUS_PRNG_SALT);
        for (int i = 0; i < bladeCount; i++)
        {
            double r = prng.Uniform(0.0, 1.0);
            if (r >= Constants.CACTUS_PROBABILITY) continue;

            double height = prng.Uniform(Constants.CACTUS_HEIGHT_MIN, Constants.CACTUS_HEIGHT_MAX);
            double width = prng.Uniform(Constants.CACTUS_WIDTH_MIN, Constants.CACTUS_WIDTH_MAX);
            double armDraw = prng.Uniform(0.0, 1.0);
            double noArmThreshold = 1.0 - Constants.CACTUS_ARM_PROBABILITY;
            double twoArmThreshold = noArmThreshold + Constants.CACTUS_TWO_ARM_PROBABILITY * Constants.CACTUS_ARM_PROBABILITY;
            if (armDraw < noArmThreshold)
            {
                return new ExpectedCactus(i, 0, height, width, +1);
            }
            if (armDraw < twoArmThreshold)
            {
                return new ExpectedCactus(i, 2, height, width, +1);
            }

            sbyte side = prng.Uniform(0.0, 1.0) < 0.5 ? (sbyte)-1 : (sbyte)+1;
            return new ExpectedCactus(i, 1, height, width, side);
        }

        throw new InvalidOperationException("canonical seed produced no cactus slot");
    }

    private static int ExpectedTumbleweedCount(double monitorWidth)
    {
        if (monitorWidth < 480.0) return 0;
        int count = (int)Math.Floor(monitorWidth / 1920.0 * Constants.TUMBLEWEED_COUNT_PER_1920DIP);
        return count < 1 ? 1 : count;
    }

    [Fact]
    public void DesertConstantsArePinned()
    {
        Assert.Equal(0.005, Constants.CACTUS_PROBABILITY);
        Assert.Equal(30.0, Constants.CACTUS_HEIGHT_MIN);
        Assert.Equal(70.0, Constants.CACTUS_HEIGHT_MAX);
        Assert.Equal(0xFF2D7A2Du, Constants.CACTUS_COLOR);
        Assert.Equal(4, Constants.TUMBLEWEED_COUNT_PER_1920DIP);
        Assert.Equal(120.0, Constants.TUMBLEWEED_SPEED_MAX);
        Assert.Equal(0x7B0117CA7B0117CAul, Constants.TUMBLEWEED_PRNG_SALT);
    }

    [Fact]
    public void SetSceneDesertClearsEntitiesAndGeneratesCacti()
    {
        var sim = BuildSim();
        sim.Entities.Add(new Entity { Kind = EntityKind.Snowflake });

        sim.SetScene(Scene.Desert);

        Assert.Equal(Scene.Desert, sim.CurrentScene);
        Assert.Equal(ExpectedTumbleweedCount(Monitor1920), sim.Entities.Count);
        Assert.All(sim.Entities, e => Assert.Equal(EntityKind.Tumbleweed, e.Kind));

        int cactusCount = sim.Blades.Count(b => b.IsCactus);
        Assert.InRange(cactusCount, 1, 10);
    }

    [Fact]
    public void FirstCactusMatchesSpecDerivedPrngSnapshot()
    {
        var sim = BuildSim();
        var expected = FirstExpectedCactus(sim.Blades.Length);

        sim.SetScene(Scene.Desert);

        Blade b = sim.Blades[expected.SlotIndex];
        Assert.True(b.IsCactus);
        Assert.Equal(expected.Type, b.CactusType);
        Assert.Equal(expected.Height, b.CactusHeight, 12);
        Assert.Equal(expected.Width, b.CactusWidth, 12);
        if (expected.Type == 1) Assert.Equal(expected.ArmSide, b.CactusArmSide);
    }

    [Fact]
    public void GrassSceneRestoresOriginalFlowerAndMushroomSlotVariants()
    {
        var sim = BuildSim();
        var expected = FirstExpectedCactus(sim.Blades.Length);

        sim.Blades[expected.SlotIndex].IsFlower = true;
        sim.Blades[expected.SlotIndex].IsMushroom = true;
        sim.Blades[expected.SlotIndex].OriginalIsFlower = true;
        sim.Blades[expected.SlotIndex].OriginalIsMushroom = true;

        sim.SetScene(Scene.Desert);
        Assert.True(sim.Blades[expected.SlotIndex].IsCactus);
        Assert.False(sim.Blades[expected.SlotIndex].IsFlower);
        Assert.False(sim.Blades[expected.SlotIndex].IsMushroom);

        sim.SetScene(Scene.Grass);
        Assert.False(sim.Blades[expected.SlotIndex].IsCactus);
        Assert.True(sim.Blades[expected.SlotIndex].IsFlower);
        Assert.True(sim.Blades[expected.SlotIndex].IsMushroom);
    }

    [Fact]
    public void DesertGeneratesExpectedTumbleweedCount()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Desert);

        int expected = ExpectedTumbleweedCount(Monitor1920);
        Assert.True(expected >= 1);
        Assert.Equal(expected, sim.Entities.Count);
    }

    [Fact]
    public void FirstTumbleweedMatchesSpecDerivedPrngSnapshot()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Desert);
        Assert.NotEmpty(sim.Entities);

        var prng = Prng.Init(Constants.CANONICAL_TEST_SEED ^ Constants.TUMBLEWEED_PRNG_SALT);
        double expectedSize = prng.Uniform(Constants.TUMBLEWEED_SIZE_MIN, Constants.TUMBLEWEED_SIZE_MAX);
        double expectedX = prng.Uniform(0.0, Monitor1920);
        double expectedY = sim.GroundY - prng.Uniform(Constants.TUMBLEWEED_Y_OFFSET_MIN, Constants.TUMBLEWEED_Y_OFFSET_MAX);
        double speed = prng.Uniform(Constants.TUMBLEWEED_SPEED_MIN, Constants.TUMBLEWEED_SPEED_MAX);
        double direction = prng.Uniform(0.0, 1.0) < 0.5 ? -1.0 : 1.0;
        double expectedVx = direction * speed;
        double expectedRotation = prng.Uniform(0.0, 2.0 * Math.PI);

        Entity e = sim.Entities[0];
        Assert.Equal(EntityKind.Tumbleweed, e.Kind);
        Assert.Equal(expectedSize, e.Size, 12);
        Assert.Equal(expectedX, e.X, 12);
        Assert.Equal(expectedY, e.Y, 12);
        Assert.Equal(expectedVx, e.Vx, 12);
        Assert.Equal(expectedRotation, e.Rotation, 12);
        Assert.Equal(expectedVx / expectedSize, e.RotationSpeed, 12);
    }

    [Fact]
    public void TumbleweedRespawnsAtOppositeEdgeWhenOffScreen()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Desert);
        Assert.NotEmpty(sim.Entities);

        Entity e = sim.Entities[0];
        e.X = sim.MonitorWidth + 100.0;
        sim.Entities[0] = e;

        sim.TickEntities(0.0);

        e = sim.Entities[0];
        Assert.Equal(EntityKind.Tumbleweed, e.Kind);
        Assert.Equal(-e.Size, e.X, 12);
        Assert.True(e.Vx > 0.0);
    }

    [Fact]
    public void DesertSceneLeavesCanonicalFirstBladeGeometryBitIdentical()
    {
        var sim = BuildSim(density: 1.0);
        sim.SetScene(Scene.Desert);

        Blade first = sim.Blades[0];
        Assert.Equal(4.941073726820111, first.BaseX, 12);
        Assert.Equal(24.469991818248864, first.Height, 12);
        Assert.Equal(1.5829214329729786, first.Thickness, 12);
        Assert.Equal((byte)3, first.Hue);
        Assert.Equal(3.3176304956845826, first.SwayPhaseOffset, 12);
        Assert.Equal(0.97444439458772458, first.Stiffness, 12);
    }
}

File: tests/DesktopGrass.Native.Tests/src/desert_tests.cpp
// desert_tests.cpp - §14 Desert scene cacti + tumbleweeds.

#include "TestHelpers.h"
#include "Sim.h"
#include "snapshot_data.h"

#include <cmath>
#include <cstddef>
#include <vector>

using namespace desktopgrass;

namespace {

constexpr double kMonitor1920 = 1920.0;

struct ExpectedCactus {
    std::size_t slotIndex = 0;
    uint8_t type = 0;
    double height = 0.0;
    double width = 0.0;
    int8_t armSide = +1;
};

ExpectedCactus first_expected_cactus(std::size_t bladeCount) {
    Prng p;
    prng_init(p, CANONICAL_TEST_SEED ^ CACTUS_PRNG_SALT);

    for (std::size_t i = 0; i < bladeCount; ++i) {
        const double r = prng_uniform(p, 0.0, 1.0);
        if (r >= CACTUS_PROBABILITY) continue;

        ExpectedCactus expected{};
        expected.slotIndex = i;
        expected.height = prng_uniform(p, CACTUS_HEIGHT_MIN, CACTUS_HEIGHT_MAX);
        expected.width = prng_uniform(p, CACTUS_WIDTH_MIN, CACTUS_WIDTH_MAX);

        const double armDraw = prng_uniform(p, 0.0, 1.0);
        const double noArmThreshold = 1.0 - CACTUS_ARM_PROBABILITY;
        const double twoArmThreshold = noArmThreshold + CACTUS_TWO_ARM_PROBABILITY * CACTUS_ARM_PROBABILITY;
        if (armDraw < noArmThreshold) {
            expected.type = 0;
        } else if (armDraw < twoArmThreshold) {
            expected.type = 2;
        } else {
            expected.type = 1;
            expected.armSide = prng_uniform(p, 0.0, 1.0) < 0.5
                ? static_cast<int8_t>(-1)
                : static_cast<int8_t>(+1);
        }
        if (expected.height < CACTUS_ARM_MIN_HEIGHT) {
            expected.type = 0;
            expected.armSide = +1;
        }
        return expected;
    }

    Assert::Fail(L"Canonical seed produced no cactus slot", LINE_INFO());
    return {};
}

int expected_tumbleweed_count(double monitorWidth) {
    if (monitorWidth < 480.0) return 0;
    int count = static_cast<int>(std::floor(monitorWidth / 1920.0 * static_cast<double>(TUMBLEWEED_COUNT_PER_1920DIP)));
    return count < 1 ? 1 : count;
}

} // anonymous

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace DesktopGrassNativeTests
{
TEST_CLASS(DesertTests)
{
public:
TEST_METHOD(DesertConstantsArePinned) {
    Assert::IsTrue(CACTUS_PROBABILITY == Near(0.005));
    Assert::IsTrue(CACTUS_HEIGHT_MIN == Near(30.0));
    Assert::IsTrue(CACTUS_HEIGHT_MAX == Near(70.0));
    Assert::IsTrue(CACTUS_COLOR == 0xFF2D7A2Du);
    Assert::IsTrue(TUMBLEWEED_COUNT_PER_1920DIP == 4);
    Assert::IsTrue(TUMBLEWEED_SPEED_MAX == Near(72.0));
    Assert::IsTrue(TUMBLEWEED_PRNG_SALT == 0x7B0117CA7B0117CAull);
}

TEST_METHOD(SimSetSceneDesertClearsEntitiesAndGeneratesCacti) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    Entity fake{};
    fake.kind = EntityKind::Snowflake;
    sim.entities.push_back(fake);

    sim_set_scene(sim, Scene::Desert);

    Assert::IsTrue(sim.currentScene == Scene::Desert);
    Assert::IsTrue(sim.entities.size() == static_cast<std::size_t>(expected_tumbleweed_count(kMonitor1920)));
    for (const Entity& e : sim.entities) Assert::IsTrue(e.kind == EntityKind::Tumbleweed);

    std::size_t cactusCount = 0;
    for (const Blade& b : sim.blades) if (b.isCactus) ++cactusCount;
    Assert::IsTrue(cactusCount >= 1);
    Assert::IsTrue(cactusCount <= 10);
}

TEST_METHOD(FirstCactusMatchesTheSpecDerivedPRNGSnapshot) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    const ExpectedCactus expected = first_expected_cactus(sim.blades.size());

    sim_set_scene(sim, Scene::Desert);

    Assert::IsTrue(expected.slotIndex < sim.blades.size());
    const Blade& b = sim.blades[expected.slotIndex];
    Assert::IsTrue(b.isCactus);
    Assert::IsTrue(b.cactusType == expected.type);
    Assert::IsTrue(b.cactusHeight == Near(expected.height).margin(1e-12));
    Assert::IsTrue(b.cactusWidth == Near(expected.width).margin(1e-12));
    if (expected.type == 1) Assert::IsTrue(b.cactusArmSide == expected.armSide);
}

TEST_METHOD(GrassSceneRestoresOriginalFlowerAndMushroomSlotVariants) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    const ExpectedCactus expected = first_expected_cactus(sim.blades.size());
    Assert::IsTrue(expected.slotIndex < sim.blades.size());

    Blade& target = sim.blades[expected.slotIndex];
    target.isFlower = true;
    target.isMushroom = true;
    target.originalIsFlower = true;
    target.originalIsMushroom = true;

    sim_set_scene(sim, Scene::Desert);
    Assert::IsTrue(sim.blades[expected.slotIndex].isCactus);
    Assert::IsFalse(sim.blades[expected.slotIndex].isFlower);
    Assert::IsFalse(sim.blades[expected.slotIndex].isMushroom);

    sim_set_scene(sim, Scene::Grass);
    Assert::IsFalse(sim.blades[expected.slotIndex].isCactus);
    Assert::IsTrue(sim.blades[expected.slotIndex].isFlower);
    Assert::IsTrue(sim.blades[expected.slotIndex].isMushroom);
}

TEST_METHOD(DesertGeneratesTheExpectedTumbleweedCount) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    sim_set_scene(sim, Scene::Desert);

    const int expected = expected_tumbleweed_count(kMonitor1920);
    Assert::IsTrue(expected >= 1);
    Assert::IsTrue(sim.entities.size() == static_cast<std::size_t>(expected));
}

TEST_METHOD(FirstTumbleweedMatchesTheSpecDerivedPRNGSnapshot) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    sim_set_scene(sim, Scene::Desert);
    Assert::IsFalse(sim.entities.empty());

    Prng p;
    prng_init(p, CANONICAL_TEST_SEED ^ TUMBLEWEED_PRNG_SALT);
    const double expectedSize = prng_uniform(p, TUMBLEWEED_SIZE_MIN, TUMBLEWEED_SIZE_MAX);
    const double expectedX = prng_uniform(p, 0.0, kMonitor1920);
    const double expectedY = sim.windowHeight - prng_uniform(p, TUMBLEWEED_Y_OFFSET_MIN, TUMBLEWEED_Y_OFFSET_MAX);
    const double speed = prng_uniform(p, TUMBLEWEED_SPEED_MIN, TUMBLEWEED_SPEED_MAX);
    const double direction = prng_uniform(p, 0.0, 1.0) < 0.5 ? -1.0 : 1.0;
    const double expectedVx = direction * speed;
    const double expectedRotation = prng_uniform(p, 0.0, 6.28318530717958647692);

    const Entity& e = sim.entities[0];
    Assert::IsTrue(e.kind == EntityKind::Tumbleweed);
    Assert::IsTrue(e.size == Near(expectedSize).margin(1e-12));
    Assert::IsTrue(e.x == Near(expectedX).margin(1e-12));
    Assert::IsTrue(e.y == Near(expectedY).margin(1e-12));
    Assert::IsTrue(e.vx == Near(expectedVx).margin(1e-12));
    Assert::IsTrue(e.rotation == Near(expectedRotation).margin(1e-12));
    Assert::IsTrue(e.rotationSpeed == Near(expectedVx / expectedSize).margin(1e-12));
}

TEST_METHOD(TumbleweedRespawnsAtTheOppositeEdgeWhenOffScreen) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    sim_set_scene(sim, Scene::Desert);
    Assert::IsFalse(sim.entities.empty());

    sim.entities[0].x = sim.monitorWidth + 100.0;
    sim_tick_entities(sim, 0.0);

    const Entity& e = sim.entities[0];
    Assert::IsTrue(e.kind == EntityKind::Tumbleweed);
    Assert::IsTrue(e.x == Near(-e.size).margin(1e-12));
    Assert::IsTrue(e.vx > 0.0);
}

TEST_METHOD(TumbleweedHopsAboveItsBaselineThenSettles) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    sim_set_scene(sim, Scene::Desert);
    Assert::IsFalse(sim.entities.empty());

    double yBase = sim.entities[0].altitudeAnchor;
    Assert::IsTrue(sim.entities[0].y == Near(yBase).margin(1e-9)); // starts grounded

    double minY = sim.entities[0].y;
    for (int i = 0; i < 900; ++i) {
        sim_tick_entities(sim, 1.0 / 60.0);
        Entity& t = sim.entities[0];
        // Pin x on-screen so it doesn't roll off and respawn mid-test.
        if (t.x < 50.0) t.x = 50.0;
        if (t.x > sim.monitorWidth - 50.0) t.x = sim.monitorWidth - 50.0;
        yBase = t.altitudeAnchor;
        minY = std::min(minY, t.y);
        Assert::IsTrue(t.y <= yBase + 1e-6); // never sinks below the baseline
    }

    Assert::IsTrue(minY < yBase - 1.0); // it left the ground at least once
}

TEST_METHOD(DesertSceneLeavesTheCanonicalFirstBladeGeometryBitIdentical) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, 1.0);
    Assert::IsTrue(sim.blades.size() == desktopgrass::test::CANONICAL_BLADE_COUNT);

    sim_set_scene(sim, Scene::Desert);

    const Blade& first = sim.blades[0];
    const auto& expected = desktopgrass::test::CANONICAL_FIRST_10[0];
    Assert::IsTrue(first.baseX == Near(expected.baseX).margin(1e-12));
    Assert::IsTrue(first.height == Near(expected.height).margin(1e-12));
    Assert::IsTrue(first.thickness == Near(expected.thickness).margin(1e-12));
    Assert::IsTrue(first.hue == expected.hue);
    Assert::IsTrue(first.swayPhaseOffset == Near(expected.sway).margin(1e-12));
    Assert::IsTrue(first.stiffness == Near(expected.stiffness).margin(1e-12));
}
};
}
